data_structures/priority_queue.cpp

```diff
@@ -0,0 +1,170 @@
+/**
+ * @file
+ * @brief Implemetation of Max Priority Queue uning max-heap
+ * More about it: https://en.wikipedia.org/wiki/Priority_queue
+ * @details
+ * In computer science, a priority queue is an abstract data-type similar to a 
+ * regular queue or stack data structure in which each element additionally has a 
+ * priority associated with it. In a priority queue, an element with high priority 
+ * is served before an element with low priority.
+ * Though being a relatively new data structure it has its use cases in lots of fields:
+ * Data compression[Huffman codes], Graph Searching[Dijkstra's Algorithm], 
+ * Artificial Intelligence[A* seach], etc.
+ * Insertion - O(log(N))
+ * Deletion - O(log(N))
+ * Max - O(1)
+ * All the method names have been directly taken from PQ implementation of Java
+ * @author [Puneet Tripathi](https://github.com/PuneetTri)
+ */
+
+#include<iostream> // For standard I/O
+#include<cassert> // For assert
+#include<utility> // For importing swap utility function
+
+// Class 
+class PQ{
+    private:
+        int capacity; // max capacity of PQ
+        int *arr; // Pointer to the array of the PQ
+        int N; // current size of PQ or total elements currently in PQ
+
+        /* Helper function for the max-heap implementation of PQ
+        *  Function would help promte child node to their correct 
+        *  position
+        */
+        void swim(int pos) {
+            // Check if parent is smaller than child node
+            while(pos > 1 && (arr[pos/2] < arr[pos])) {
+                // In such case swap value of child with parent
+                std::swap(arr[pos], arr[pos/2]);
+                pos = pos/2; // Jump to position of parent node
+            }
+            // Promotion of child node will go on until it becomes smaller than the parent
+        }
+
+        /* Helper function for the max-heap implementation of PQ
+        *  Function would help demote parent node to their correct 
+        *  position
+        */
+        void sink(int pos) {
```

Comment:
Functions should be documented explaining the use of its parameters, a brief description, and what it returns.